Formula/ruby@1.8.7.rb

```diff
@@ -0,0 +1,205 @@
+
+  resource "rubygems" do
+    url "https://rubygems.org/rubygems/rubygems-2.5.0.tgz"
+    sha256 "a39e95acb86f52fdca38b1484b2eed61a7f76551119079d7a9f8c67c5be78831"
+  end
+
+  def install
+    # Compilation with `superenv` breaks because the Ruby build system sets
+    # `RUBYLIB` and `RUBYOPT` and this disturbs the wrappers for `cc` and
+    # `clang` that are provided by Homebrew and are implemented in Ruby. We
+    # prefix those invocations with `env` to clean the environment for us.
+    scrub_env = "/usr/bin/env RUBYLIB= RUBYOPT="
+    inreplace "mkconfig.rb", "    if /^prefix$/ =~ name\n",
+              <<-EOS.undent.gsub(/^/, "    ")
+                # `env` removes stuff that will break `superenv`.
+                if %w[CC CPP LDSHARED LIBRUBY_LDSHARED].include?(name)
+                  val = val.sub("\\"", "\\"#{scrub_env} ")
+                end
+                if /^prefix$/ =~ name
+              EOS
+
+    args = %W[
+      --prefix=#{prefix}
+      --enable-shared
+      --with-sitedir=#{HOMEBREW_PREFIX}/lib/ruby/site_ruby
+      --with-vendordir=#{HOMEBREW_PREFIX}/lib/ruby/vendor_ruby
+    ]
+
+    if build.universal?
+      ENV.universal_binary
+      args << "--with-arch=#{Hardware::CPU.universal_archs.join(",")}"
+    end
+
+    args << "--program-suffix=#{program_suffix}" if build.with? "suffix"
+    args << "--enable-install-doc" if build.with? "doc"
+
+    rm_r "ext/tk" if build.without? "tcltk"
+
+    system "./configure", *args
+
+    # Ruby has been configured to look in the HOMEBREW_PREFIX for the
+    # sitedir and vendordir directories; however we don't actually want to create
+    # them during the install.
+    #
+    # These directories are empty on install; sitedir is used for non-rubygems
+    # third party libraries, and vendordir is used for packager-provided libraries.
+    inreplace "instruby.rb" do |s|
+      s.gsub! "\n    makedirs [archlibdir, sitearchlibdir, vendorarchlibdir]\n",
+              "\n    makedirs [archlibdir]\n"
+      s.gsub! "\n    makedirs [rubylibdir, sitelibdir, vendorlibdir]\n",
+              "\n    makedirs [rubylibdir]\n"
+    end
+
+    system "make"
+    system "make", "install"
+    system "make", "install-doc" if build.with? "doc"
+
+    # Install rubygems into its own prefix, then move into the right place, as
+    # we cannot easily make it install where it should be and the defaults are
+    # no good and would cause a sandbox violation.
+    resource("rubygems").stage do
+      system "#{bin}/ruby#{program_suffix}", "setup.rb", "--prefix=#{prefix}/gem"
+      mv (prefix/"gem/bin").children, bin
+      mv (prefix/"gem/lib").children, lib/"ruby/#{abi_version}"
+      rm_r prefix/"gem"
+    end
+  end
+
+  def post_install
+    # Customize rubygems to look/install in the global gem directory
+    # instead of in the Cellar, making gems last across reinstalls
+    config_file = lib/"ruby/#{abi_version}/rubygems/defaults/operating_system.rb"
```

Comment:
I have dig in the git log, the original intent in homebrew-version is to have each specific formula to be mostly similar.  So you have bits of codes that are essentially the same, and one can just change a few methods to vary per version.

One can try a `diff ruby187.rb ruby193.rb` or `diff ruby20.rb ruby22.rb`.  With the use of a variable the diff is rather small.

homebrew-php has a nice shared abstraction that is reused by formulae. Maybe we can head to ward that later on.